engineio/src/asynchronous/client/async_client.rs

```diff
@@ -0,0 +1,335 @@
+use crate::{asynchronous::async_socket::Socket as InnerSocket, error::Result, Packet, PacketId};
+use bytes::Bytes;
+
+#[derive(Clone, Debug)]
+pub struct Client {
+    pub(super) socket: InnerSocket,
+}
+
+impl Client {
+    pub async fn close(&self) -> Result<()> {
+        self.socket.disconnect().await
+    }
+
+    /// Opens the connection to a specified server. The first Pong packet is sent
+    /// to the server to trigger the Ping-cycle.
+    pub async fn connect(&self) -> Result<()> {
+        self.socket.connect().await
+    }
+
+    /// Disconnects the connection.
+    pub async fn disconnect(&self) -> Result<()> {
+        self.socket.disconnect().await
+    }
+
+    /// Sends a packet to the server.
+    pub async fn emit(&self, packet: Packet) -> Result<()> {
+        self.socket.emit(packet).await
+    }
+
+    /// Polls for next payload
+    #[doc(hidden)]
+    pub async fn poll(&self) -> Result<Option<Packet>> {
+        let packet = self.socket.poll().await?;
+        if let Some(packet) = packet {
+            // check for the appropriate action or callback
+            self.socket.handle_packet(packet.clone()).await;
+            match packet.packet_id {
+                PacketId::MessageBinary => {
+                    self.socket.handle_data(packet.data.clone()).await;
+                }
+                PacketId::Message => {
+                    self.socket.handle_data(packet.data.clone()).await;
+                }
+                PacketId::Close => {
+                    self.socket.handle_close().await;
+                }
+                PacketId::Open => {
+                    unreachable!("Won't happen as we open the connection beforehand");
+                }
+                PacketId::Upgrade => {
+                    // this is already checked during the handshake, so just do nothing here
+                }
+                PacketId::Ping => {
+                    self.socket.pinged().await?;
+                    self.emit(Packet::new(PacketId::Pong, Bytes::new())).await?;
+                }
+                PacketId::Pong => {
+                    // this will never happen as the pong packet is
+                    // only sent by the client
+                    unreachable!();
```

Comment:
nit: May be better to throw an error here rather than [panicing](https://doc.rust-lang.org/src/core/macros/mod.rs.html#594